fs/xfstests.py

```diff
@@ -0,0 +1,191 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+#
+# Copyright: 2016 IBM
+# Author: Praveen K Pandey <praveen@linux.vnet.ibm.com>
+#
+# Based on code by Cleber Rosa <crosa@redhat.com>
+#   copyright: 2011 Redhat
+#   https://github.com/autotest/autotest-client-tests/tree/master/xfstests
+
+
+import os
+import glob
+import re
+import shutil
+
+from avocado import Test
+from avocado import main
+from avocado.utils import process, build, git, distro
+from avocado.utils.software_manager import SoftwareManager
+
+
+class Xfstests(Test):
+
+    PASSED_RE = re.compile(r'Passed all \d+ tests')
+    FAILED_RE = re.compile(r'Failed \d+ of \d+ tests')
+    NA_RE = re.compile(r'Passed all 0 tests')
+    NA_DETAIL_RE = re.compile(r'(\d{3})\s*(\[not run\])\s*(.*)')
+    GROUP_TEST_LINE_RE = re.compile('(\d{3})\s(.*)')
+
+    def _get_available_tests(self):
+        os.chdir(self.srcdir)
+
+        tests = glob.glob(self.srcdir + '/tests/*/???.out')
+        tests += glob.glob(self.srcdir + '/tests/*/???.out.linux')
+        tests = [t.replace('.linux', '') for t in tests]
+        tests_list = [t[-7:-4] for t in tests if os.path.exists(t[:-4])]
+        tests_list.append('')
+        tests_list.sort()
+        return tests_list
+
+    def _run_sub_test(self, test):
+
+        os.chdir(self.srcdir)
+        self.log.info("Running test: %s" % test)
+        output = process.system_output('./check %s' % test,
+                                       ignore_status=True)
+        lines = output.split('\n')
+        result_line = lines[-3]
+        if self.NA_RE.match(result_line):
+            detail_line = lines[-3]
+            match = self.NA_DETAIL_RE.match(detail_line)
+            if match is not None:
+                error_msg = match.groups()[2]
+            else:
+                error_msg = 'Test dependency failed, test not run'
+            raise self.error(error_msg)
+
+        elif self.FAILED_RE.match(result_line):
+            raise self.error('Test error, check debug logs for complete '
+                             'test output')
+
+        elif self.PASSED_RE.match(result_line):
+            return
+
+        else:
+            raise self.error('Could not assert test success or failure, '
+                             'assuming failure. Please check debug logs')
+
+    def _get_groups(self):
+        '''
+        Returns the list of groups known to xfstests
+        By reading the group file and identifying unique mentions of groups
+        '''
+        groups = []
+        for l in open(os.path.join(self.srcdir, 'group')).readlines():
+            m = self.GROUP_TEST_LINE_RE.match(l)
+            if m is not None:
+                groups = m.groups()[1].split()
+                for g in groups:
+                    if g not in groups:
+                        groups.add(g)
```

Comment:
groups is a list... `add()` is not available for lists. You should use `append()` here.